lib/weakhash/weakhash.go

```diff
@@ -8,111 +8,125 @@ package weakhash
 
 import (
 	"bufio"
-	"context"
 	"io"
 
 	"github.com/chmduquesne/rollinghash/adler32"
 )
 
-const (
-	Size = 4
+// A Finder scans through an io.ReaderAt, looking for blocks whose Adler-32
+// hash is in a given set.
+type Finder struct {
+	buf    []byte
+	err    error
+	hashes map[uint32]struct{}
+	offset int64
 
-	// don't track more hits than this for any given weakhash
-	maxWeakhashFinderHits = 10
-)
+	hf *adler32.Adler32
 
-// Find finds all the blocks of the given size within io.Reader that matches
-// the hashes provided, and returns a hash -> slice of offsets within reader
-// map, that produces the same weak hash.
-func Find(ctx context.Context, ir io.Reader, hashesToFind []uint32, size int) (map[uint32][]int64, error) {
-	if ir == nil || len(hashesToFind) == 0 {
-		return nil, nil
-	}
+	r    io.ReaderAt
+	sr   *io.SectionReader
+	br   *bufio.Reader
+	size int64 // File size.
 
-	r := bufio.NewReader(ir)
-	hf := adler32.New()
+	// Last matching hash value.
+	match uint32
+}
 
-	n, err := io.CopyN(hf, r, int64(size))
-	if err == io.EOF {
-		return nil, nil
-	}
-	if err != nil {
-		return nil, err
-	}
-	if n != int64(size) {
-		return nil, io.ErrShortBuffer
-	}
+const int64Max = 1<<63 - 1
 
-	offsets := make(map[uint32][]int64)
-	for _, hashToFind := range hashesToFind {
-		offsets[hashToFind] = make([]int64, 0, maxWeakhashFinderHits)
+// NewFinder returns a Finder that initially has no hashes.
+// Call Add before Next to add hashes.
+//
+// The buffer buf will be filled for each match found. Its length is taken
+// to be the block size.
+func NewFinder(r io.ReaderAt, buf []byte) *Finder {
+	rr, ok := r.(io.Reader)
```

Comment:
We should just change the interface to be what it needs to be

Reply:
Done.